tools/lemmacmd.go

```diff
@@ -0,0 +1,267 @@
+			os.Exit(255)
+		}
+
+		secretService, err := secret.New(&secret.Config{KeyBytes: key})
+		if err != nil {
+			fmt.Printf("lemmacmd: unable to build lemma.secret: %v\n", err)
+			os.Exit(255)
+		}
+
+		plaintextBytes, err := ioutil.ReadFile(*inputpath)
+		if err != nil {
+			fmt.Printf("lemmacmd: unable to read plaintext file %q: %v\n", *inputpath, err)
+			os.Exit(255)
+		}
+
+		sealedData, err := secretService.Seal(plaintextBytes)
+		if err != nil {
+			fmt.Printf("lemmacmd: unable to seal plaintext: %v\n", err)
+			os.Exit(255)
+		}
+
+		err = writeCiphertext(keySalt, *itercount, sealedData, *outputpath)
+		if err != nil {
+			fmt.Printf("lemmacmd: unable to write sealed data to disk: %v\n", err)
+			os.Exit(255)
+		}
+	case "dec":
+		keySalt, keyIter, sealedData, err := readCiphertext(*inputpath)
+		if err != nil {
+			fmt.Printf("lemmacmd: unable to read ciphertext file: %v\n", err)
+			os.Exit(255)
+		}
+
+		key, err := generateDecryptionKey(keySalt, keyIter, *keypath)
+		if err != nil {
+			fmt.Printf("lemmacmd: unable to build key: %v\n", err)
+			os.Exit(255)
+		}
+
+		secretService, err := secret.New(&secret.Config{KeyBytes: key})
+		if err != nil {
+			fmt.Printf("lemmacmd: unable to build lemma.secret: %v\n", err)
+			os.Exit(255)
+		}
+
+		plaintextBytes, err := secretService.Open(sealedData)
+		if err != nil {
+			fmt.Printf("lemmacmd: unable to open ciphertext: %v\n", err)
+			os.Exit(255)
+		}
+
+		err = ioutil.WriteFile(*outputpath, plaintextBytes, 0600)
+		if err != nil {
+			fmt.Printf("lemmacmd: unable to write plaintext bytes to disk: %v\n", err)
+			os.Exit(255)
+		}
+	default:
+		fmt.Printf("lemmacmd: invalid mode provided: %v\n", *mode)
+		os.Exit(255)
+	}
+}
+
+// Returns encryption key either from a 32-byte base64 encoded key on disk or from a key
+// derivation function which uses a passphrase and a randomly generated salt as input.
+func generateEncryptionKey(keypath string, keyiter int) (keySalt []byte, key *[secret.SecretKeyLength]byte, err error) {
+	// if a keypath is given try and use it
+	if keypath != "" {
+		key, err := secret.ReadKeyFromDisk(keypath)
+		if err != nil {
+			return nil, nil, fmt.Errorf("unable to build secret service: %v", err)
+		}
+		return nil, key, nil
+	}
+
+	// otherwise read in a passphrase from disk and use that, remember to reset your terminal afterwards
+	var passphrase string
+	fmt.Printf("Passphrase: ")
+	fmt.Scanln(&passphrase)
+
+	// derive key and return it
+	keySalt, key, err = deriveKeyFromPassphrase(keyiter, passphrase)
+	if err != nil {
+		return nil, nil, fmt.Errorf("unable to derive key from passphrase: %v", err)
+	}
+
+	return keySalt, key, nil
+}
+
+// Returns encryption key either from a 32-byte base64 encoded key on disk or from a key
+// derivation function which uses a passphrase and salt as input.
+func generateDecryptionKey(salt []byte, keyiter int, keypath string) (key *[secret.SecretKeyLength]byte, err error) {
```

Comment:
`generateEncryptionKey` and `generateDecryptionKey` are pretty similar. Merge them? https://github.com/mailgun/lemma/commit/a9c6939033c41a20e9549b540a89e4ce21157430